### PR TITLE
[Internal] Change Feed: Adds emulator tests

### LINKED ISSUE #2194: [Track] Enable emulator test after release.
SDK PR: https://github.com/Azure/azure-cosmos-dotnet-v3/pull/2193
I will make an update here when released.

### PR DESCRIPTION
**Pull request by zishanfu333**
# [Internal] Change Feed: Adds emulator tests

## Description
Add tests for emulator change feed fix. 

1. Validates Incremental Change Feed by inserting and deleting documents and verifying nothing reported
2. Validates error message with Full Fidelity Change Feed and start from beginning.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #2194

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 1]
```diff
@@ -660,6 +660,42 @@ public async Task ChangeFeedIteratorCore_CannotMixTokensFromOtherContainers()
...
+            // Insert documents and then delete them
+            int totalDocuments = 50;
+            IList<ToDoActivity> createdItems = await this.CreateRandomItems(container, totalDocuments, randomPartitionKey: true);
+            foreach (ToDoActivity item in createdItems)
+            {
+                await container.DeleteItemAsync<ToDoActivity>(item.id, new PartitionKey(item.pk));
+            }
+            FeedIterator<ToDoActivityWithMetadata> fullFidelityIterator = container.GetChangeFeedIterator<ToDoActivityWithMetadata>(
+                ChangeFeedStartFrom.Beginning(),
+                ChangeFeedMode.Incremental);
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:684]: This is incremental and not full fidelity even though the variable name is `fullFidelityIterator`

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 2]
```diff
@@ -660,6 +660,42 @@ public async Task ChangeFeedIteratorCore_CannotMixTokensFromOtherContainers()
...
+            while (fullFidelityIterator.HasMoreResults)
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:685]: No need for this loop if you expect the 304 on the first call.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 3]
```diff
@@ -734,6 +770,39 @@ public async Task ChangeFeedIteratorCore_WithFullFidelity()
...
+                cancellationToken: this.cancellationToken);
+            ContainerInternal container = (ContainerInternal)response;
+            int totalDocuments = 10;
+            await this.CreateRandomItems(container, totalDocuments, randomPartitionKey: true);
+
+            // FF does not work with StartFromBeginning currently, capture error
+            FeedIterator<ToDoActivityWithMetadata> fullFidelityIterator = container.GetChangeFeedIterator<ToDoActivityWithMetadata>(
+                ChangeFeedStartFrom.Beginning(),
+                ChangeFeedMode.FullFidelity);
+            while (fullFidelityIterator.HasMoreResults)
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:792]: No need for the loop.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 4]
```diff
@@ -734,6 +770,39 @@ public async Task ChangeFeedIteratorCore_WithFullFidelity()
...
+            ContainerProperties properties = new ContainerProperties(id: Guid.NewGuid().ToString(), partitionKeyPath: ChangeFeedIteratorCoreTests.PartitionKey);
+            properties.ChangeFeedPolicy.FullFidelityRetention = TimeSpan.FromMinutes(5);
+            ContainerResponse response = await this.database.CreateContainerAsync(
+                properties,
+                cancellationToken: this.cancellationToken);
+            ContainerInternal container = (ContainerInternal)response;
+            int totalDocuments = 10;
+            await this.CreateRandomItems(container, totalDocuments, randomPartitionKey: true);
+
+            // FF does not work with StartFromBeginning currently, capture error
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:788]: Is this temporary or by design? If there are some combinations that are not allowed, then we need to show that in the type system for the parameters for GetChangeFeedIterator.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs (hunk 1 above)
Comment (zishanfu333) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:685]: Nice catch. Thanks!

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs (hunk 4 above)
Comment (zishanfu333) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:780]: Yes, it is temporary.

**Comment by zishanfu333** [comment]
Since there is a fix for the emulator and the check-in system might need to get the new version, I marked the test as ignore for now.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 5]
```diff
@@ -660,6 +660,33 @@ public async Task ChangeFeedIteratorCore_CannotMixTokensFromOtherContainers()
             Assert.AreEqual(HttpStatusCode.BadRequest, responseMessage.StatusCode);
         }
 
+        /// <summary>
+        /// This test validates Incremental Change Feed by inserting and deleting documents and verifying nothing reported
+        /// </summary>
+        [Ignore]
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:666]: Why is the test being  ignored?

**Comment by j82w** [comment]
> Since there is a fix for the emulator and the check-in system might need to get the new version, I marked the test as ignore for now.

There is an estimate on when the new emulator will be released? Can we create a tracking item somewhere so the ignore gets cleaned up in the future.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 6]
```diff
@@ -660,6 +660,33 @@ public async Task ChangeFeedIteratorCore_CannotMixTokensFromOtherContainers()
...
+        [TestMethod]
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:667]: @zishanfu333 Please add an Issue to track enabling this test so we don't forget, or we can wait until the new Emulator is released to merge this PR, not sure what is the timeline

Comment (zishanfu333) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:667]: Issue: https://github.com/Azure/azure-cosmos-dotnet-v3/issues/2194
I will make an update once it released. Thanks!

**Review by ealsur — APPROVED** [review]
Thanks for adding the tracking item

**Review by j82w — APPROVED** [review]

**Review by ealsur — APPROVED** [review]

**Pull request merged**